src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

```diff
@@ -249,16 +249,39 @@ async function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourc
   const existingPostRun = moduleConfig.postRun || [];
   (moduleConfig as any).preloadPlugins = [];
 
+  var totalResources = Object.keys(resources.assembly).length + Object.keys(resources.pdb || {}).length + /* dotnet.wasm */ 1;
+
+  if (resourceLoader.bootConfig.resources.runtime.hasOwnProperty('dotnet.timezones.blat')){
+    totalResources++;
+  }
+
+  if (resourceLoader.bootConfig.icuDataMode !== ICUDataMode.Invariant){
+    totalResources++;
+  }
+
+  var resourcesLoaded = 0;
+  function setProgress(){
+    resourcesLoaded++;
+    const targetElement = document.getElementById('app') as HTMLDivElement;
+    const percentage = resourcesLoaded / totalResources * 100;
+    targetElement.style.setProperty('--blazor-load-percentage', percentage + '%');
+    document.getElementById('progress')!.style.display = 'block';
+    document.getElementById('percentage')!.textContent = Math.floor(percentage) + '%';
```

Comment:
In terms of performance, given this may be called a couple hundred times, would it make sense to cache the elements so we aren't calling `getElementById` repeatedly? Granted, the benefit from caching here is likely to be pretty small.

Reply:
Also, I don't believe the `not null` (`!`) assertion is safe for these elements. Users may remove those elements from the `index.html`, or they may be on a project created using a prior version of the template which didn't have these elements.

Can we possibly add `null` checks for `app`, `progress`, `percentage`. We may be able to short-circuit out of this `setProgress` if those elements don't exist. 

In prior versions of this template, the `app` was an element tag instead of a `div` with an `app` `id`.
https://github.com/dotnet/aspnetcore/blob/592dfe165b6fccb65accd20c3581a430b987127d/src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Client/wwwroot/index.html#L18

Reply:
I have added code to cache the elements and additional checks to make sure that these elements are not null before their style is set.

When changing `<div id = 'app'>` to `<app>`, there was an exception when debugging using developer tools, as it was looking for #app. I have changed it back to an `app` `id` to see if this will work.

Reply:
> When changing `<div id = 'app'>` to `<app>`, there was an exception when debugging using developer tools, as it was looking for #app. I have changed it back to an `app` `id` to see if this will work.

Thanks for looking into this! I brought up the `<app>` tag simply as a reference to how we did this before to ensure we don't unduly break existing apps. Just to clarify, this isn't something we'd actually want to change in our current template.